addons/product/models/product_attribute.py

```diff
@@ -134,14 +136,56 @@ def write(self, values):
             self.invalidate_cache()
         return res
 
+    def _get_related_variants(self):
+        """Returns all related archived products.
+
+        When product is referenced from a SO, the variant is archived instead
+        of unlinked when it becomes not necessary. Thus, it can happen
+        that a product.attribute.value is still referenced by an archived
+        product.
+        """
+        self.ensure_one()
+        query = """
+            SELECT array_agg(pp.id)
+            FROM product_product pp
+            JOIN product_variant_combination pvc ON pp.id = pvc.product_product_id 
+            JOIN product_template_attribute_value ptav ON pvc.product_template_attribute_value_id = ptav.id
+            JOIN product_attribute_value pav ON pav.id = ptav.product_attribute_value_id
+            WHERE pav.id = %(value_id)s;
+        """
+        self.env.cr.execute(query, {"value_id": self.id})
+        # result is `[(None,)]` if nothing matched, `[([ids],)]` otherwise
+        product_ids = self.env.cr.fetchall()[0][0]
+        return self.env["product.product"].browse(product_ids)
+
+    def _archive(self):
+        return self.write({"active": False})
+
+    def _get_pav_to_archive(self):
+        pav_to_archive = self.env[self._name]
```

Comment:
Use a set or a list instead of a recordset, it is faster:
https://gist.github.com/guewen/a4a2eba5d128b4ed3129866f80571fbd
```suggestion
        pav_to_archive_ids = set()
```